app/status.go

```diff
@@ -315,6 +315,34 @@ func (a *App) SetStatusDoNotDisturb(userId string) {
 	a.Publish(event)
 }
 
+func (a *App) SetStatusOutOfOffice(userId string) {
+	if !*a.Config().ServiceSettings.EnableUserStatuses {
+		return
+	}
+
+	status, err := a.GetStatus(userId)
+
+	if err != nil {
+		status = &model.Status{UserId: userId, Status: model.STATUS_OUT_OF_OFFICE, Manual: false, LastActivityAt: 0, ActiveChannel: ""}
+	}
+
+	status.Status = model.STATUS_OUT_OF_OFFICE
+	status.Manual = true
+
+	a.AddStatusCache(status)
+
+	if result := <-a.Srv.Store.Status().SaveOrUpdate(status); result.Err != nil {
+		l4g.Error(utils.T("api.status.save_status.error"), userId, result.Err)
+	}
+
+	event := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_STATUS_CHANGE, "", "", status.UserId, nil)
+	event.Add("status", model.STATUS_OUT_OF_OFFICE)
+	event.Add("user_id", status.UserId)
+	a.Go(func() {
+		a.Publish(event)
+	})
```

Comment:
`Publish` calls shouldn't be made in new goroutines like this anymore (see https://github.com/mattermost/mattermost-server/pull/8373).